hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieIndexer.java

```diff
@@ -0,0 +1,292 @@
+          return -1;
+        }
+      }
+    }, "Indexer failed");
+  }
+
+  @TestOnly
+  public Option<String> doSchedule() throws Exception {
+    return this.scheduleIndexing(jsc);
+  }
+
+  private Option<String> scheduleIndexing(JavaSparkContext jsc) throws Exception {
+    String schemaStr = UtilHelpers.getSchemaFromLatestInstant(metaClient);
+    try (SparkRDDWriteClient<HoodieRecordPayload> client = UtilHelpers.createHoodieClient(jsc, cfg.basePath, schemaStr, cfg.parallelism, Option.empty(), props)) {
+      return doSchedule(client);
+    }
+  }
+
+  private Option<String> doSchedule(SparkRDDWriteClient<HoodieRecordPayload> client) {
+    List<MetadataPartitionType> partitionTypes = getRequestedPartitionTypes(cfg.indexTypes);
+    Option<String> indexingInstant = client.scheduleIndexing(partitionTypes);
+    if (!indexingInstant.isPresent()) {
+      LOG.error("Scheduling of index action did not return any instant.");
+    }
+    return indexingInstant;
+  }
+
+  private int runIndexing(JavaSparkContext jsc) throws Exception {
+    String schemaStr = UtilHelpers.getSchemaFromLatestInstant(metaClient);
+    try (SparkRDDWriteClient<HoodieRecordPayload> client = UtilHelpers.createHoodieClient(jsc, cfg.basePath, schemaStr, cfg.parallelism, Option.empty(), props)) {
+      if (isNullOrEmpty(cfg.indexInstantTime)) {
+        // Instant time is not specified
+        // Find the earliest scheduled indexing instant for execution
+        Option<HoodieInstant> earliestPendingIndexInstant = metaClient.getActiveTimeline()
+            .filterPendingIndexTimeline()
+            .filter(i -> !(i.isCompleted() || INFLIGHT.equals(i.getState())))
+            .firstInstant();
+        if (earliestPendingIndexInstant.isPresent()) {
+          cfg.indexInstantTime = earliestPendingIndexInstant.get().getTimestamp();
+          LOG.info("Found the earliest scheduled indexing instant which will be executed: "
+              + cfg.indexInstantTime);
+        } else {
+          throw new HoodieIndexException("There is no scheduled indexing in the table.");
+        }
+      }
+      return handleResponse(client.index(cfg.indexInstantTime)) ? 0 : 1;
+    }
+  }
+
+  private int scheduleAndRunIndexing(JavaSparkContext jsc) throws Exception {
+    String schemaStr = UtilHelpers.getSchemaFromLatestInstant(metaClient);
+    try (SparkRDDWriteClient<HoodieRecordPayload> client = UtilHelpers.createHoodieClient(jsc, cfg.basePath, schemaStr, cfg.parallelism, Option.empty(), props)) {
+      Option<String> indexingInstantTime = doSchedule(client);
+      if (indexingInstantTime.isPresent()) {
+        return handleResponse(client.index(indexingInstantTime.get())) ? 0 : 1;
+      } else {
+        return -1;
+      }
+    }
+  }
+
+  private int dropIndex(JavaSparkContext jsc) throws Exception {
+    List<MetadataPartitionType> partitionTypes = getRequestedPartitionTypes(cfg.indexTypes);
```

Comment:
makes sense. 
Currently, bloom_filters partition is built only for record key so we have just one partition, so this works. However, i get your meta point. As suggested earlier, i think it would be better to refactor indexing APIs in a way such that it only asks what users would expect and then construct the configs, partition paths, etc.

Reply:
I think its better to do it more streamlined to begin with. Redoing/reworking adds a lot of overhead to understand side effects